File: dailyCodingProblems/js/twoNumberSum.js
/*
Problem #1 [Easy]
Given a list of numbers and a number k, return whether any two numbers from the list add up to k.

For example, given [10, 15, 3, 7] and k of 17, return true since 10 + 7 is 17.

Bonus: Can you do this in one pass?
*/

// time: O(N) || space: O(N)
function twoNumberSum(numbers, k) {
  let processedNumbers = {};

  for(let i = 0; i < numbers.length; i++) {
    let current = numbers[i];
    let complement = k - current;
    if (complement in processedNumbers) {
      return true;
    }

    processedNumbers[current] = true;
  }

  return false;
}

File: dailyCodingProblems/java/TwoNumberSum.java
package dailyCodingProblems.java;

import java.util.HashMap;

public class TwoNumberSum {
  static boolean twoNumberSum(int[] numbers, int k) {
    HashMap<Integer, Boolean> processedNumbers = new HashMap<>();

    for (int i = 0; i < numbers.length; i++) {
      int current = numbers[i];
      int complement = k - current;
      if (processedNumbers.containsKey(complement)) {
        return true;
      }

      processedNumbers.put(current, true);
    }

    return false;
  }

  public static void main(String[] args) {
    int[] numbers = {10, 15, 3, 7};
    int k = 17;

    System.out.println(twoNumberSum(numbers, k));
  }
}
